sweep/fee_bumper.go

```diff
@@ -518,6 +519,207 @@ type monitorRecord struct {
+		// Check all monitored txns to see if any of them needs to be
+		// bumped.
+		t.processRecords()
+
+	case <-t.quit:
+		log.Debug("Fee bumper stopped, exit monitor")
+		return
+	}
+}
+
+// processRecords checks all the txns being monitored, and check if any of them
+// needs to be bumped. If so, it will attempt to bump the fee of the tx.
+func (t *TxPublisher) processRecords() {
+	// confirmed is a helper closure that returns a boolean indicating
+	// whether the tx is confirmed.
+	confirmed := func(txid chainhash.Hash,
+		event *chainntnfs.ConfirmationEvent) bool {
+
+		select {
+		case event := <-event.Confirmed:
+			log.Debugf("Sweep tx %v confirmed in block %v", txid,
+				event.BlockHeight)
+
+			return true
+
+		default:
+			return false
+		}
+	}
+
+	// visitor is a helper closure that visits each record and performs an
+	// RBF if necessary.
+	visitor := func(requestID uint64, r *monitorRecord) error {
+		oldTxid := r.tx.TxHash()
+		log.Tracef("Checking monitor record=%v for tx=%v", requestID,
+			oldTxid)
+
+		// If the tx is already confirmed, we can stop monitoring it.
+		if confirmed(oldTxid, r.confEvent) {
+			// Create a result that will be sent to the resultChan
+			// which is listened by the caller.
+			result := &BumpResult{
+				CurrentTx: r.tx,
+				Confirmed: true,
+				requestID: requestID,
+			}
+
+			// Notify that this tx is confirmed and remove the
+			// record from the map.
+			t.notifyAndRemove(result)
+
+			// Move to the next record.
+			return nil
+		}
+
+		// Get the current conf target for this record.
+		confTarget := t.calcCurrentConfTarget(r.set.DeadlineHeight())
+
+		// Ask the fee function whether a bump is needed. We expect the
+		// fee function to increase its returned fee rate after calling
+		// this method.
+		if r.feeFunction.SkipFeeBump(confTarget) {
+			log.Debugf("Skip bumping tx %v at height=%v", oldTxid,
+				t.currentHeight)
+
+			return nil
+		}
+
+		// The fee function now has a new fee rate, we will use it to
+		// bump the fee of the tx.
+		result, err := t.createAndPublishTx(requestID, r)
+		if err != nil {
+			log.Errorf("Failed to bump tx %v: %v", oldTxid, err)
+
+			// Return nil so the visitor can continue to the next
+			// record.
+			return nil
+		}
+
+		// Notify the new result.
+		t.notifyAndRemove(result)
+
+		return nil
+	}
+
+	// TODO(yy): need to double check as the above `visitor` will put data
+	// in the sync map, so we may need to do a read first.
+	t.records.ForEach(visitor)
+}
+
+// createAndPublishTx creates a new tx with a higher fee rate and publishes it
+// to the network. It will update the record with the new tx and fee rate if
+// successfully created, and return the result when published successfully.
+func (t *TxPublisher) createAndPublishTx(requestID uint64,
+	r *monitorRecord) (*BumpResult, error) {
+
+	// Fetch the old tx.
+	oldTx := r.tx
+
+	// Create a new tx with the new fee rate.
+	tx, fee, err := t.createAndCheckTx(r.set, r.changeScript, r.feeFunction)
```

Comment:
Maybe add a more detailed comment why when reaching this point the FeeRate is already increased?

Reply:
Done.